### PR TITLE
Added brute-force protection

### PR DESCRIPTION
**Pull request by KizerovDmitriy**
# GreenCityUser PR
Added brute-force protection to application.

- Integrated a new service to track and block accounts when a user enters incorrect credentials beyond the allowed number of attempts. This service helps to prevent brute-force attacks by monitoring and limiting login attempts.
- Implemented CAPTCHA validation to detect bots and prevent automated login attempts. If the CAPTCHA is invalid after a set number of tries, the account is blocked.
- Created an email template that will be sent to the user when their account is blocked, providing instructions on how to reset their password and unlock their account.
- Added unit tests to cover the brute-force protection functionality, including tracking, blocking, and unblocking logic.
- Introduced new environment variables to configure the maximum number of allowed attempts (maxAttempts) and the block duration (blockTime), allowing dynamic adjustment of security settings.
- Modified the signIn method to incorporate the new logic for checking account statuses and updating login attempts.

## Issue Link :clipboard:
[#7583](https://github.com/ita-social-projects/GreenCity/issues/7583)

# PR Checklist Forms

_(to be filled out by PR submitter)_
- [x] Code is up-to-date with the `dev` branch.
- [x] You've successfully built and run the tests locally.
- [x] There are new or updated unit tests validating the change.
- [x] JIRA/ Github Issue number & title in PR title (ISSUE-XXXX: Ticket title)
- [x] This template filled (above this section).
- [x] Sonar's report does not contain bugs, vulnerabilities, security issues, code smells ar duplication
- [x] `NEED_REVIEW` and `READY_FOR_REVIEW` labels are added.
- [x] All files reviewed before sending to reviewers

### CODE REVIEW

**Review by IliaWithHat — APPROVED** [review]

**Review by lukovskyj — APPROVED** [review]

File: core/src/main/java/greencity/security/listener/AuthenticationFailureListener.java [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+ */
+@Component
+@RequiredArgsConstructor
+public class AuthenticationFailureListener implements ApplicationListener<AuthenticationFailureBadCredentialsEvent> {
+    private final HttpServletRequest request;
+    private final LoginAttemptService loginAttemptService;
+
+    @Override
+    public void onApplicationEvent(@NotNull AuthenticationFailureBadCredentialsEvent event) {
+        final String xfHeader = request.getHeader("X-Forwarded-For");
```
Comment (RomaMocherniuk) [inline core/src/main/java/greencity/security/listener/AuthenticationFailureListener.java]: I see you use `X-Forwarded-For` several times
Please move it to a constant

File: core/src/main/java/greencity/security/listener/AuthenticationFailureListener.java [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+        if (xfHeader == null || xfHeader.isEmpty() || !xfHeader.contains(request.getRemoteAddr())) {
```
Comment (RomaMocherniuk) [inline core/src/main/java/greencity/security/listener/AuthenticationFailureListener.java]: please use  [StringUrils](https://commons.apache.org/proper/commons-lang/javadocs/api-2.6/org/apache/commons/lang/StringUtils.html#isBlank%28java.lang.String%29) `isEmpty`

File: core/src/main/java/greencity/security/listener/AuthenticationFailureListener.java [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+            loginAttemptService.loginFailed(request.getRemoteAddr());
+        } else {
+            loginAttemptService.loginFailed(xfHeader.split(",")[0]);
```
Comment (RomaMocherniuk) [inline core/src/main/java/greencity/security/listener/AuthenticationFailureListener.java]: please check the size to prevent `IndexOutOfBoundsException`

File: core/src/main/java/greencity/security/listener/AuthenticationFailureListener.java [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+ * delegates work of tracking of login attempts to {@link LoginAttemptService}.
+ */
+@Component
+@RequiredArgsConstructor
+public class AuthenticationFailureListener implements ApplicationListener<AuthenticationFailureBadCredentialsEvent> {
+    private final HttpServletRequest request;
+    private final LoginAttemptService loginAttemptService;
+
+    @Override
+    public void onApplicationEvent(@NotNull AuthenticationFailureBadCredentialsEvent event) {
```
Comment (RomaMocherniuk) [inline core/src/main/java/greencity/security/listener/AuthenticationFailureListener.java]: Why is `event` marked as `@NotNull` if you do not use it?

File: service-api/pom.xml [hunk 5]
```diff
@@ -156,6 +157,11 @@
             <version>2.2.19</version>
             <scope>compile</scope>
         </dependency>
+        <dependency>
+            <groupId>com.google.guava</groupId>
+            <artifactId>guava</artifactId>
+            <version>${guava.version}</version>
```
Comment (RomaMocherniuk) [inline service-api/pom.xml:163]: I do not see usage of this dependency
Why did you add it?

File: service-api/src/main/java/greencity/constant/ErrorMessage.java [hunk 6]
```diff
@@ -54,4 +54,6 @@ public class ErrorMessage {
     public static final String YOU_DO_NOT_HAVE_PERMISSIONS_TO_DEACTIVATE_THIS_USER =
         "You do not have permission to deactivate this user";
     public static final String BASE64_DECODE_MESSAGE = "Can't decode from base64 format";
+    public static final String BRUTEFORCE_PROTECTION_MESSAGE =
+        "User account is blocked due to too many failed login attempts";
```
Comment (RomaMocherniuk) [inline service-api/src/main/java/greencity/constant/ErrorMessage.java]: Should we add info about blocking time? Something like this:
`User account is blocked due to too many failed login attempts. Try again in an hour`

File: service/src/main/java/greencity/security/service/LoginAttemptServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+@Service
+public class LoginAttemptServiceImpl implements LoginAttemptService {
+    private final LoadingCache<String, Integer> attemptsCache;
+    private final HttpServletRequest request;
+    @Value("${bruteForceSettings.maxAttempts}")
+    private int maxAttempt;
+
+    public LoginAttemptServiceImpl(HttpServletRequest request,
+        @Value("${bruteForceSettings.blockTimeInHours}") long blockTimeInHours) {
+        super();
```
Comment (RomaMocherniuk) [inline service/src/main/java/greencity/security/service/LoginAttemptServiceImpl.java]: what is the reason for calling `super()` here?

File: service/src/main/java/greencity/security/service/LoginAttemptServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+     */
+    public void loginFailed(final String key) {
+        int attempts;
+        try {
+            attempts = attemptsCache.get(key);
+        } catch (final ExecutionException e) {
+            attempts = 0;
+        }
+        attempts++;
+        attemptsCache.put(key, attempts);
```
Comment (RomaMocherniuk) [inline service/src/main/java/greencity/security/service/LoginAttemptServiceImpl.java]: you can use [asMap](https://guava.dev/releases/19.0/api/docs/com/google/common/cache/LoadingCache.html) to simplify the logic
```
        attemptsCache.asMap().merge(key, 1, Integer::sum);

```
Please check if we have to catch any exception here

File: service/src/main/java/greencity/security/service/LoginAttemptServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+            return attemptsCache.get(getClientIP()) >= maxAttempt;
+        } catch (final ExecutionException e) {
+            return false;
+        }
+    }
+
+    private String getClientIP() {
+        final String xfHeader = request.getHeader("X-Forwarded-For");
+        if (xfHeader != null) {
+            return xfHeader.split(",")[0];
```
Comment (RomaMocherniuk) [inline service/src/main/java/greencity/security/service/LoginAttemptServiceImpl.java]: please check size to prevent `IndexOutOfBoundException`

File: service/src/main/java/greencity/security/service/OwnSecurityServiceImpl.java [hunk 10]
```diff
@@ -208,6 +221,15 @@ public SuccessSignInDto signIn(final OwnSignInDto dto) {
         return new SuccessSignInDto(user.getId(), accessToken, refreshToken, user.getName(), true);
     }
 
+    private void publishFailureEvent(String email, String password, String errorMessage) {
+        UsernamePasswordAuthenticationToken auth =
+            new UsernamePasswordAuthenticationToken(email, password);
```
Comment (RomaMocherniuk) [inline service/src/main/java/greencity/security/service/OwnSecurityServiceImpl.java]: Why do you need to publish an event with a user password?
It is sensitive information and should not be logged

**Comment by RomaMocherniuk** [comment]
Please increase code coverage
image

File: service-api/pom.xml (hunk 5 above)
Comment (KizerovDmitriy) [reply service-api/pom.xml:163]: We are using this dependency in the LoginAttemptServiceImpl class. It is declared as a private final field LoadingCache<String, Integer> and initialized in the constructor.

File: service/src/main/java/greencity/security/service/LoginAttemptServiceImpl.java (hunk 7 above)
Comment (KizerovDmitriy) [reply service/src/main/java/greencity/security/service/LoginAttemptServiceImpl.java]: I agree, it makes no sense
I fixed that

File: service-api/src/main/java/greencity/constant/ErrorMessage.java [hunk 11]
```diff
@@ -55,5 +55,5 @@ public class ErrorMessage {
         "You do not have permission to deactivate this user";
     public static final String BASE64_DECODE_MESSAGE = "Can't decode from base64 format";
     public static final String BRUTEFORCE_PROTECTION_MESSAGE =
-        "User account is blocked due to too many failed login attempts";
+        "User account is blocked due to too many failed login attempts. Try again in an hour";
```
Comment (RomaMocherniuk) [inline service-api/src/main/java/greencity/constant/ErrorMessage.java]: I see that the default value is 1 hour
`bruteForceSettings.blockTimeInHours=${BLOCK_TIME_IN_HOURS:1}`
Do we use other values?
if yes we should build a message based on them or check when the blocking time ends

File: service/src/main/java/greencity/security/service/LoginAttemptServiceImpl.java [hunk 12]
```diff
@@ -33,20 +36,15 @@ public Integer load(final String key) {
...
     public void loginFailed(final String key) {
-        int attempts;
-        try {
-            attempts = attemptsCache.get(key);
-        } catch (final ExecutionException e) {
-            attempts = 0;
-        }
-        attempts++;
-        attemptsCache.put(key, attempts);
+        attemptsCache.asMap().merge(key, 1, Integer::sum);
```
Comment (RomaMocherniuk) [inline service/src/main/java/greencity/security/service/LoginAttemptServiceImpl.java]: ~~how do we remove it after `blockTimeInHours` ends?~~
I see it, sorry
` .expireAfterWrite(blockTimeInHours, TimeUnit.HOURS)`

File: service-api/src/main/java/greencity/constant/ErrorMessage.java (hunk 11 above)
Comment (RomaMocherniuk) [reply service-api/src/main/java/greencity/constant/ErrorMessage.java]: I see that logic is based on user IP, not name
Should we modify a message to represent current logic?

File: service/src/main/java/greencity/security/service/LoginAttemptServiceImpl.java (hunk 12 above)
Comment (KizerovDmitriy) [reply service/src/main/java/greencity/security/service/LoginAttemptServiceImpl.java:41]: good remark, now I will think about how to implement it

File: service/src/main/java/greencity/security/service/LoginAttemptServiceImpl.java [hunk 13]
```diff
@@ -18,9 +22,8 @@ public class LoginAttemptServiceImpl implements LoginAttemptService {
 
     public LoginAttemptServiceImpl(HttpServletRequest request,
         @Value("${bruteForceSettings.blockTimeInHours}") long blockTimeInHours) {
-        super();
         this.request = request;
-        attemptsCache = CacheBuilder.newBuilder()
+        this.attemptsCache = CacheBuilder.newBuilder()
```
Comment (RomaMocherniuk) [inline service/src/main/java/greencity/security/service/LoginAttemptServiceImpl.java]: we use for storing unsuccess events only in-memory cache
So it seems like after redeploying the app all data will be lost. Is it for the product owners?

**Review by RomaMocherniuk — APPROVED** [review]
Left some minor comments

File: core/pom.xml [hunk 14]
```diff
@@ -183,7 +183,24 @@
...
         </dependency>
 
+        <dependency>
+            <groupId>org.springframework.cloud</groupId>
+            <artifactId>spring-cloud-starter-openfeign</artifactId>
+            <version>4.1.3</version>
+        </dependency>
+
     </dependencies>
+    <dependencyManagement>
```
Comment (RomaMocherniuk) [inline core/pom.xml]: What is the reason for using `dependencyManagement` only for this dependency?

File: core/src/main/resources/messages_en.properties [hunk 15]
```diff
@@ -42,6 +42,15 @@ sincerely.yours.greenCity=Sincerely yours, Green City team.
 sincerely.yours.Ubs=Sincerely yours, Pick Up City team.
 unsubscribe.text=If you no longer wish to receive these emails, you can
 unsubscribe=unsubscribe
+unlock=Unlock account
+block.account=Account blocked
+text.account.ban=Your account has been locked for security reasons due to a possible hacking attempt. For your safety, we recommend changing your password after restoring your account.
+warning=Warning!
+block.user=Your account is blocked
```
Comment (RomaMocherniuk) [inline core/src/main/resources/messages_en.properties:49]: why do we need `Account blocked` and `Your account is blocked`?

File: core/src/main/resources/messages_uk.properties [hunk 16]
```diff
@@ -42,6 +42,15 @@ sincerely.yours.greenCity=\u0429\u0438\u0440\u043e\u0020\u0432\u0430\u0448\u0430
...
+advice.for.block=\u0412\u0438\u0020\u043E\u0442\u0440\u0438\u043C\u0430\u043B\u0438\u0020\u0446\u0435\u0439\u0020\u0435\u043B\u0435\u043A\u0442\u0440\u043E\u043D\u043D\u0438\u0439\u0020\u043B\u0438\u0441\u0442\u0020\u0437\u0020\u043C\u0456\u0440\u043A\u0443\u0432\u0430\u043D\u044C\u0020\u0431\u0435\u0437\u043F\u0435\u043A\u0438\u002E\u0020\u0429\u043E\u0431\u0020\u0437\u0430\u0445\u0438\u0441\u0442\u0438\u0442\u0438\u0020\u0441\u0432\u0456\u0439\u0020\u043E\u0431\u043B\u0456\u043A\u043E\u0432\u0438\u0439\u0020\u0437\u0430\u043F\u0438\u0441\u002C\u0020\u0440\u0435\u043A\u043E\u043C\u0435\u043D\u0434\u0443\u0454\u043C\u043E\u0020\u0437\u043C\u0456\u043D\u0438\u0442\u0438\u0020\u043F\u0430\u0440\u043E\u043B\u044C\u0020\u043F\u0456\u0441\u043B\u044F\u0020\u0440\u043E\u0437\u0431\u043B\u043E\u043A\u0443\u0432\u0430\u043D\u043D\u044F\u0020\u043E\u0431\u043B\u0456\u043A\u043E\u0432\u043E\u0433\u043E\u0020\u0437\u0430\u043F\u0438\u0441\u0443\u002E
```
Comment (RomaMocherniuk) [inline core/src/main/resources/messages_uk.properties:53]: Why don't we use Ukrainian here?

File: service-api/pom.xml [hunk 17]
```diff
@@ -156,7 +157,27 @@
...
+        </dependency>
+        <dependency>
+            <groupId>org.springframework.cloud</groupId>
+            <artifactId>spring-cloud-starter-openfeign</artifactId>
+            <version>4.1.3</version>
+        </dependency>
     </dependencies>
 
-
+    <dependencyManagement>
```
Comment (RomaMocherniuk) [inline service-api/pom.xml]: Any reason for using `dependencyManagement` only for this dependency?

File: core/pom.xml (hunk 14 above)
Comment (KizerovDmitriy) [reply core/pom.xml]: My bad, i'll refactored this!

File: service-api/pom.xml (hunk 17 above)
Comment (KizerovDmitriy) [reply service-api/pom.xml]: My bad, i'll refactored this!

File: core/src/main/resources/messages_uk.properties (hunk 16 above)
Comment (KizerovDmitriy) [reply core/src/main/resources/messages_uk.properties:53]: We use the translation, it is a little upper in code

File: core/src/main/resources/messages_en.properties (hunk 15 above)
Comment (KizerovDmitriy) [reply core/src/main/resources/messages_en.properties:49]: 'Account blocked' used here : '<title th:text="#{block.account}"></title>'

**Review by K4sik — APPROVED** [review]

**Review by holotsvan — APPROVED** [review]

**Review by urio999 — APPROVED** [review]
